File: src/main/java/seremis/geninfusion/handler/GIEventHandler.java
package seremis.geninfusion.handler;


import cpw.mods.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.event.entity.EntityJoinWorldEvent;
import net.minecraftforge.event.world.WorldEvent;
import seremis.geninfusion.lib.DefaultProps;
import seremis.geninfusion.soul.entity.EntitySoulCustom;
import seremis.geninfusion.soul.entity.EntitySoulCustomCreature;
import seremis.geninfusion.world.GIWorldSavedData;

public class GIEventHandler {

    @SubscribeEvent
    public void joinWorld(EntityJoinWorldEvent event) {
        if(event.entity instanceof EntitySoulCustom) {
            System.out.println(event.entity + " server? " + !event.world.isRemote);
        } else if(event.entity instanceof EntitySoulCustomCreature) {
            System.out.println(event.entity + " server? " + !event.world.isRemote);
        }
    }

    @SubscribeEvent
    public void loadWorld(WorldEvent.Load event) {
        event.world.perWorldStorage.setData(DefaultProps.ID(), new GIWorldSavedData());
    }
}

File: src/main/scala/com/seremis/geninfusion/handler/GIEventHandler.scala
package com.seremis.geninfusion.handler

import com.seremis.geninfusion.api.GIApiInterface
import com.seremis.geninfusion.api.genetics.ISoul
import com.seremis.geninfusion.api.lib.Genes
import com.seremis.geninfusion.api.util.VariableName
import com.seremis.geninfusion.soulentity.SoulEntityLiving
import com.seremis.geninfusion.soulentity.logic.FieldLogic
import net.minecraft.entity.monster.EntityZombie
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.EnumHand
import net.minecraftforge.event.entity.player.PlayerInteractEvent.RightClickBlock
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent

class GIEventHandler {


    @SubscribeEvent
    def playerInteract(event: RightClickBlock) {
        if(!event.getWorld.isRemote && event.getEntityPlayer.getHeldItem(EnumHand.MAIN_HAND) == null) {

            //val method = FunctionLib.FuncEntityEntityInit

            //println(method)

           // val m = GIApiInterface.entityMethodRegistry.getMethodsForName(method)

           // println(m)

            val entity = new SoulEntityLiving(event.getWorld, GIApiInterface.geneDefaultsRegistry.getSoulForClass(classOf[EntityZombie]))

            entity.setPosition(event.getHitVec.xCoord, event.getHitVec.yCoord + 1, event.getHitVec.zCoord)

            println(entity)

            entity.getEntityId


            val geneName = Genes.GeneTest

            println(geneName)


            val compound = new NBTTagCompound

            entity.setVar(VariableName("abcdef", classOf[Int]), 100)
            entity.makePersistent(VariableName("abcdef", classOf[Int]))

            entity.asInstanceOf[SoulEntityLiving].fieldLogic.writeToNBT(compound)

            println(compound)

            val logicOut = new FieldLogic(entity)

            logicOut.readFromNBT(compound)

            println(logicOut.getVar(VariableName("abcdef", classOf[Int])))

            println(GIApiInterface.geneDefaultsRegistry.getDefaultValueForClass(classOf[EntityZombie], geneName))

            val soul = GIApiInterface.geneDefaultsRegistry.getSoulForClass(classOf[EntityZombie])
            println(soul)

            val nbt = new NBTTagCompound

            GIApiInterface.dataTypeRegistry.writeValueToNBT(nbt, "soul", classOf[ISoul], soul)

            println(nbt)

            val out2 = GIApiInterface.dataTypeRegistry.readValueFromNBT(nbt, "soul", classOf[ISoul])

            println(out2)

            println(soul == out2)
        }
    }
}
